[PATCH v6] api: crypto: add compl queue capability

asasidharan opened the pull request.

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2446,6 +2459,37 @@ static void crypto_test_check_alg_sha512(void)
+	qparam.sched.sync  = ODP_SCHED_SYNC_PARALLEL;
+	qparam.sched.group = ODP_SCHED_GROUP_ALL;
+
+	out_queue = odp_queue_create("crypto-out", &qparam);
```
out_queue is not needed, just return odp_queue_create()

asasidharan replied:
Ack.

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2457,14 +2501,47 @@ static int crypto_suite_sync_init(void)
+		fprintf(stderr, "Crypto outq creation failed.\n");
+		return -1;
+	}
+	suite_context.queue = odp_queue_lookup("crypto-out");
```
Instead of lookup, just save the out_queue handle here.

asasidharan replied:
Ack.

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2457,14 +2501,47 @@ static int crypto_suite_sync_init(void)
+		fprintf(stderr, "Crypto outq creation failed.\n");
+		return -1;
+	}
 	suite_context.queue = odp_queue_lookup("crypto-out");
```
Instead of lookup, just save the out_queue handle here.

asasidharan replied:
Ack.

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2483,21 +2560,71 @@ static int crypto_suite_packet_sync_init(void)
+		fprintf(stderr, "Crypto outq creation failed.\n");
+		return -1;
+	}
 	suite_context.queue = odp_queue_lookup("crypto-out");
```
Instead of lookup, just save the out_queue handle here.

asasidharan replied:
Ack.

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2483,21 +2560,71 @@ static int crypto_suite_packet_sync_init(void)
+		fprintf(stderr, "Crypto outq creation failed.\n");
+		return -1;
+	}
+	suite_context.queue = odp_queue_lookup("crypto-out");
```
Instead of lookup, just save the out_queue handle here.

asasidharan replied:
Ack.

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2752,6 +2889,19 @@ static int crypto_init(odp_instance_t *inst)
+	odp_schedule_config_init(&sched_config);
+
+	/* Configure the scheduler. All test cases share the config. */
+	if (odp_schedule_config(&sched_config)) {
```
odp_schedule_config(NULL) could be used instead when all config params have default values.

asasidharan replied:
Agreed.

psavol commented on include/odp/api/spec/crypto.h:
```diff
@@ -893,6 +893,27 @@ typedef struct odp_crypto_capability_t {
+	/**
+	 * Scheduled crypto completion queue support
+	 *
+	 * This defines whether schedule queues are supported as crypto
```
typo: schedule -> scheduled

asasidharan replied:
Ack.

asasidharan commented:
Updated v2 with:
- Typo corrected in the API change.
- Removed odp_queue_lookup() calls from suite init functions.
- Removed use of unnecessary variable out_queue from sched_compl_queue_create().

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2752,6 +2876,17 @@ static int crypto_init(odp_instance_t *inst)
 		return -1;
 	}
 
+	if (odp_schedule_capability(&sched_capa)) {
```
Sched_capa call is not needed. Capa is set and not used.

It seems that odp_crypto_capability() call under could be removed also. Algorithm tests use is anyway.

asasidharan replied:
Agreed. Removed capability calls for sched and crypto from crypto_init() in v3.

JannePeltonen commented:
I am not sure if I like the proposal but I can see that a similar thing was done for timers a little while back. Would you need if for ipsec, compression and classification too?

If this API change is adopted, then also example/ipsec_crypto and test/performance/odp_crypto.c should check the new capability fields.

asasidharan commented:
> 
> 
>  Would you need if for ipsec, compression and classification too?
> 

I have plans to update the same for IPsec as well.

> If this API change is adopted, then also example/ipsec_crypto and test/performance/odp_crypto.c should check the new capability fields.

Agreed. I will update these two applications to check the new capability fields in v4 of the PR.

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2708,29 +2824,35 @@ odp_testinfo_t crypto_suite[] = {
 #define ODP_CRYPTO_PACKET_SYNC_INP  "odp_crypto_packet_sync_inp"
-#define ODP_CRYPTO_PACKET_ASYNC_INP "odp_crypto_packet_async_inp"
+#define ODP_CRYPTO_PACKET_ASYNC_PLAIN_INP "odp_crypto_packet_async_plain_inp"
+#define ODP_CRYPTO_PACKET_ASYNC_SCHED_INP "odp_crypto_packet_async_sched_inp"
```
All these defines can be removed, as only place those are used is just under here.

asasidharan replied:
Ack.  Removed the defines in v5.

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2446,6 +2459,34 @@ static void crypto_test_check_alg_sha512(void)
+
+static odp_event_t sched_compl_queue_deq(void)
+{
+	return odp_schedule(&suite_context.queue, ODP_SCHED_NO_WAIT);
```
schedule() call does not need queue parameter (should use NULL instead). Passing the suite queue handle there is risky, as on an error/bug the call could overwrite some other queue handle (or invalid) there.

asasidharan replied:
Ack. Changed the queue parameter passed to NULL in v5.

psavol commented on test/validation/api/crypto/odp_crypto_test_inp.c:
```diff
@@ -2483,21 +2553,67 @@ static int crypto_suite_packet_sync_init(void)
+{
+	odp_queue_t out_queue;
+
+	out_queue = odp_queue_lookup("crypto-out");
```
This could be simplified by using suite_context.queue instead of lookup. Each suite_init would need to initialize it to ODP_QUEUE_INVALID.

asasidharan replied:
Agreed. With v5, updated the change to use suite_context.queue instead of doing odp_queue_lookup().

psavol commented:
Needs reviewed-by tags still

asasidharan commented:
Added Reviewed-by tag for Petri and rebased to the latest master with v6. @JannePeltonen should I add your reviewed-by tag to the application changes?

JannePeltonen commented:
> should I add your reviewed-by tag to the application changes?

No need to.

asasidharan commented:
@psavol Is there anything pending from my side on this PR?

psavol commented:
> 
> 
> @psavol Is there anything pending from my side on this PR?

No. When a PR is labeled "API ready", it's ready to be merged but waits for other API PRs to get ready and to be merged together.
